File: POTD/Day-03_Flattening-a-Linked-List/Solution.cpp
/*
class Node {
public:
    int data;
    Node* next;
    Node* bottom;

    Node(int x) {
        data = x;
        next = NULL;
        bottom = NULL;
    }
};
*/

class Solution {
  private:
    // Helper function to merge two sorted bottom-linked lists
    Node* merge(Node* a, Node* b) {
        // If one list is empty, return the other
        if (a == NULL) return b;
        if (b == NULL) return a;

        Node* result = NULL;

        // Choose smaller node and recurse
        if (a->data < b->data) {
            result = a;
            result->bottom = merge(a->bottom, b);
        } else {
            result = b;
            result->bottom = merge(a, b->bottom);
        }

        // next pointer is not used in final list
        result->next = NULL;
        return result;
    }

  public:
    Node *flatten(Node *root) {
        // Base case: empty list or single column
        if (root == NULL || root->next == NULL) {
            return root;
        }

        // Step 1: Recursively flatten the list on the right
        root->next = flatten(root->next);

        // Step 2: Merge current list with flattened right list
        root = merge(root, root->next);

        return root;
    }
};

File: POTD/Day-03_Flattening-a-Linked-List/Solution.py
'''
class Node:
    def __init__(self, d):
        self.data = d
        self.next = None
        self.bottom = None
'''

class Solution:
    # Helper function to merge two sorted bottom-linked lists
    def merge(self, a, b):

        # If one list is empty, return the other
        if a is None:
            return b
        if b is None:
            return a

        # Choose smaller node and recurse
        if a.data < b.data:
            result = a
            result.bottom = self.merge(a.bottom, b)
        else:
            result = b
            result.bottom = self.merge(a, b.bottom)

        # next pointer is not used in final list
        result.next = None
        return result

    def flatten(self, root):
        # Base case: empty list or single column
        if root is None or root.next is None:
            return root

        # Step 1: Flatten the list on the right
        root.next = self.flatten(root.next)

        # Step 2: Merge current list with flattened right list
        root = self.merge(root, root.next)

        return root
